standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java

```diff
@@ -1255,21 +1256,71 @@ private void create_database_core(RawStore ms, final Database db)
         LOG.error("No such catalog " + db.getCatalogName());
         throw new InvalidObjectException("No such catalog " + db.getCatalogName());
       }
-      Path dbPath = wh.determineDatabasePath(cat, db);
+      final Path dbPath = wh.determineDatabasePath(cat, db);
+      final Path dbExternalPath = wh.getDefaultExternalDatabasePath(db.getName());
       db.setLocationUri(dbPath.toString());
 
       boolean success = false;
-      boolean madeDir = false;
+      boolean madeManagedDir = false;
+      boolean madeExternalDir = false;
       Map<String, String> transactionalListenersResponses = Collections.emptyMap();
       try {
         firePreEvent(new PreCreateDatabaseEvent(db, this));
-        if (!wh.isDir(dbPath)) {
-          LOG.debug("Creating database path " + dbPath);
-          if (!wh.mkdirs(dbPath)) {
-            throw new MetaException("Unable to create database path " + dbPath +
-                ", failed to create database " + db.getName());
+        try {
+          // Since this may be done as random user (if doAs=true) he may not have access
+          // to the managed directory. We run this as an admin user
+          madeManagedDir = UserGroupInformation.getLoginUser().doAs(
+              new PrivilegedExceptionAction<Boolean>() {
+                @Override
+                public Boolean run() throws MetaException {
+                  if (!wh.isDir(dbPath)) {
+                    LOG.info("Creating database path in managed directory " + dbPath);
+                    if (!wh.mkdirs(dbPath)) {
+                      throw new MetaException("Unable to create database managed path " + dbPath +
+                          ", failed to create database " + db.getName());
+                    }
+                    return true;
+                  }
+                  return false;
+                }
+              });
+          if (madeManagedDir) {
+            LOG.info("Created database path in managed directory " + dbPath);
+          }
+        } catch (IOException | InterruptedException e) {
+          throw new MetaException("Unable to create database managed directory " + dbPath +
+              ", failed to create database " + db.getName());
+        }
+        // We only create the external directory is the table is on the default managed warehouse directory
+        // and the external warehouse directory has been set
+        if (wh.hasExternalWarehouseRoot() &&
+            wh.isDatabaseLocationDefault(db)) {
+          try {
+            madeExternalDir = UserGroupInformation.getCurrentUser().doAs(
```

Comment:
@thejasmn, in the case of remote HiveMetastore, is UserGroupInformation.getCurrentUser() set to the user that issued the request to the Metastore, or is it still the "hive" user?